### DISCUSSION TITLE
建议将事件的'sub_type'改为可空

### DISCUSSION THREAD

**Discussion by LaoLittle** [howmanybots/onebot#224] [想法 / 建议]
sub_type不一定有值，而且他们都处于top-level，导致编写序列化时有一定的抽象复杂度

**Comment by stdrc** [comment]
现在没有值的时候是空字符串吧

**Reply by LaoLittle** [reply]
我的意思是null或直接删去

**Reply by stdrc** [reply]
好处是什么呢（

**Reply by LaoLittle** [reply]
> 好处是什么呢（

或者说去掉sub_type，直接在事件内部指定type，这东西有点阴间

**Reply by stdrc** [reply]
确实考虑在 OneBot 13 把所有 `xx_type` 统一成 `type`，用点号分割，比如 `message.private.from_group`，但 12 可能还是不动比较好

**Reply by LaoLittle** [reply]
> 确实考虑在 OneBot 13 把所有 `xx_type` 统一成 `type`，用点号分割，比如 `message.private.from_group`，但 12 可能还是不动比较好

12还是个草案，13就开始制定了么（

**Reply by LaoLittle** [reply]
还有就是所有东西都是同一级的感觉有点离谱

完全可以
- type
  - detail_type
  - data1
  - data2
- meta

**Reply by stdrc** [reply]
> 12还是个草案，13就开始制定了么（

12 准备发正式版了吧，不太想有明显改动了

**Reply by stdrc** [reply]
其实 `sub_type` 算是从酷Q继承下来的余孽了（
还存在于 OneBot 12 确实是因为当时没想到要改，但现在已经不太好大改了，下版再说吧

**Reply by LaoLittle** [reply]
> 其实 `sub_type` 算是从酷Q继承下来的余孽了（
> 还存在于 OneBot 12 确实是因为当时没想到要改，但现在已经不太好大改了，下版再说吧

期待捏

**Comment by stdrc** [comment]
我觉得允许为 `null` 会导致一些语言里面处理更麻烦

**Reply by LaoLittle** [reply]
params也是可空（null）的，这个就不影响了么（

**Reply by LaoLittle** [reply]
哦记错了，是{}

**Reply by stdrc** [reply]
没有**标准接口**允许参数值为 `null`，但确实有可以不传的，这里的逻辑是，以方便用户为先，OneBot 实现可以去复杂一点，判断某个参数存不存在

**Reply by LaoLittle** [reply]
那就空的时候删去此field，类似echo

**Reply by stdrc** [reply]
> 那就空的时候删去此field，类似echo

其实 `sub_type` 当时是考虑让所有事件的基本字段都一致，所以才用 `sub_type = ''` 来表示
